쥬스 메이커 [STEP 3] marisol, mmim

JoSH0318 opened the pull request:
- @stevenkim18  안녕하세요 스티븐! 🤗
- STEP3 PR입니다! 말씀하신대로 main branch에 merge해서 PR 합니다! 잘 부탁드립니다. 
---
### UML
<img width="5440" alt="UML (1)" src="https://user-images.githubusercontent.com/88810018/157183415-23aa128c-2936-4e25-8b47-3e1de38e2a07.png">

---
### 📌주요기능
#### class OrderViewController
- `func showCurrentStock()`: 현재 과일 재고를 `stockLabel`에 보여주는 기능
- `func presentStockViewController()`: StockViewController로 화면 전환해주는 기능
- `func orderButtonsClicked()`: IBAction Collection으로 구현, 버튼별 동작을 매칭해주는 기능 + 재고 갱신하기 위한 showCurrentStock() 호출
- `order(_ juice: Juice)`: do catch문으로 쥬스를 주문하고, 재고 부족 및 기타 오류 발생시 catch하는 기능
- `func alertOrderCompletion()`: 주문 완료시 생성되는 알럿 기능
- `func alertOutOfStock()`: 재고 부족 시 생성되는 알럿 기능
#### class StockViewController
- `func showCurrentStock()`: OrderViewController에서 받아온 현재 과일 재고를 `stockLabel`에 보여주는 기능 + stepper의 value 값을 현재 재고로 지정
- `func stepperClicked()`: IBAction Collection으로 구현, 버튼별 Label.text 갱신 + currentStock 갱신
- `func closeButtonClicked()`: 뒤로가기 + delegate가 수행할 메서드 등록 

---

### 📌고민한점과 해결방법 / 이렇게 해결한 이유

#### 데이터 전송 방식 고민 -> Delegation pattern 선택
- OrderViewController와 StockViewController가 서로 과일 재고의 데이터를 주고 받아야 했고, 데이터 전송의 방법 중 notification center와 delegation pattern을 고민해보았습니다. notification center는 특정 이벤트에 대해 알림을 요청한 구독자 모두에게 해당 이벤트가 발생하면 알림 발송해주는데, 일대일로 데이터를 주고 받는 방식에서는 쥬스메이커의 상황에서는 delegation pattern이 적절하다고 생각해서 delegation pattern을 공부해서 사용하게 되었습니다.

#### Switch문 & if문
- switch문은 모든 case를 언급하지 않으면 default 구문을 사용해줘야 합니다. 따라서 무결점한 코드가 이뤄지기 어렵다는 단점이 있습니다. 만약 enum 타입과 같이 case가 명확히 정해져있다면 switch문을 통해 각 case를 구현해준다면 무결점한 코드가 되겠지만, 그렇지 않다면 default를 언급해줘야해서 무결점하지 못하다는 것을 알게 됐습니다. 이러한 경우에는 오히려 if-else if문으로 구현하는 것이 좀 더 무결점하다는 생각을 했습니다. 그래서 OrderViewController의 orderButtonsClicked 메서드와 StockViewController의 stepperClicked 메서드에서 if-else if문을 사용해서 구현했습니다.

---

### ❓궁금한점❓

#### Dictionary Optional Binding
```swift
private func showCurrentStock() {
    currentStocks.keys.forEach { fruit in
        switch fruit {
        case .strawberry:
            strawberryStockLabel.text = currentStocks[.strawberry]?.description
            strawberryStepper.value = Double(currentStocks[.strawberry] ?? 0)
        case .banana:
            bananaStockLabel.text = currentStocks[.banana]?.description
            bananaStepper.value = Double(currentStocks[.banana] ?? 0)                 
        case .pineapple:
            pineappleStockLabel.text = currentStocks[.pineapple]?.description
            pineappleStepper.value = Double(currentStocks[.pineapple] ?? 0)           
        case .kiwi:
            kiwiStockLabel.text = currentStocks[.kiwi]?.description
            kiwiStepper.value = Double(currentStocks[.kiwi] ?? 0)
        case .mango:
            mangoStockLabel.text = currentStocks[.mango]?.description
            mangoStepper.value = Double(currentStocks[.mango] ?? 0)
        }
    }
}
```
STEP1부터 과일과 과일의 재고를 Dictionary 타입으로 관리했습니다. OrderViewController에서 StockViewController로 넘어왔을때 과일과 재고를 Label과 Stepper에 세팅해줄때, Dictionary 타입의 key들을 forEach로 돌렸습니다. forEach문을 돌면서 과일별로 현재 재고량을 stepper의 value에 넣어주는데, stepper의 타입이 Double이기 때문에 Dictionary의 value를 Double로 형변환 하고 옵셔널 바인딩을 해줘야 했습니다. 이때 guard let이나 if let으로 옵셔널 바인딩을 하면 switch 문 내부에 들여쓰기가 한번 더 추가되어서, coalesce를 사용하여 옵셔널바인딩 하는 방법을 선택했습니다. 현재로서는 Inventory라는 Dictionary의 과일 key에 대한 value가 nil이 될 가능성이 없다고 생각하는데, `?? 0`을 사용하지 않고 옵셔널 바인딩을 더 안전하게 할 수 있는 방법이 있을지 궁금합니다.

#### Alert 합치기
지난 번 코멘트에서 Alert을 한개의 함수로 합칠 수도 있을 것 같다고 언급하셨습니다. 최대한 노력해봤지만 if문으로 분기시켜주는 방법 밖에 생각나지 않았습니다. 때문에 하나로 합칠 경우 오히려 복잡해지고 코드의 양이 줄어든다는 느낌이 없더라구요... 어떻게 하면 좋은 방향으로 하나로 합칠 수 있을지 궁금합니다. 

#### Auto-Layout과 UIStepper
오토레이아웃을 적용하면서 UIStepper의 크기가 커지거나 줄어들지 않고 원래 그대로의 크기를 유지하는 것 같습니다. 원래 UIStepper 같은 경우는 크기를 커스텀할 수 없는 지 궁금하고, 만약 그렇다면 stepper를 따로 모양과 기능을 커스텀해서 사용하는지 궁금합니다.

#### Property observer
- 원래 코드는 A라는 값을 label에 보여주고자 할 때, `label.text = A`라고 직접 넣어주는 방식을 택했습니다. 하지만 A의 값이 변할 때 마다 label 값에 직접 대입해야했고, 또한 사용자에게 보여지는 기능을 담당하는 `label`에 직접 값을 저장하는 느낌이 들었습니다. 따라서 property observer를 사용해보는 게 어떨까 고민했습니다. 
```swift
let A = 10 {
    didSet {
        label.text = "\(A)"    
    }
}
```
- 이런 식으로 property observer를 설정해준다면, 주문 또는 stepper의 변화로 인해 A값의 변화를 인지하고 자동으로 label를 갱신이 가능하겠구나 생각했습니다. 하지만 A의 값이 변할때마다 `juiceMaker.fruitStore.inventory`값도 갱신시켜줘야했고, 이러한 이유로 오히려 코드가 길어지고, 복잡해진다는 것을 깨닫고 결국 철회했습니다. 
- 저희가 고민한 것처럼 Property observer을 이용하여 `Label`을 갱신시켜주는 경우가 많은지, 또는 좋은 방향성인지 궁금합니다.

marisol-develop committed:
Merge pull request #4 from JoSH0318/5_mmim90-2

5 mmim90 2

marisol-develop committed:
[feat]: configure auto layout of StackViewController

- StackViewController에 오토레이아웃 설정

JoSH0318 committed:
[feat]: delete Singleton pattern, method, modify auto layout

- 싱글톤 제거
- convertFruitStockToString 메서드 제거
- auto layout 수정

marisol-develop committed:
[feat]: add delegation pattern, IBOutlet, IBAction

- 델리게이트 패턴 일부 구현
- StockViewController의 Stepper 관련 IBOutlet, IBAction 연결

JoSH0318 committed:
[feat]: add showCurrentStock method, @IBAction method implement

- StockViewController에 라벨 갱신을 위한 showCurrentStock 메서드 추가
- StockViewController @IBAction func stepperClicked 로직 구현
- 에러 알럿 extension으로 구현

marisol-develop committed:
[improve]: change switch statement to if statement

- switch문을 if문으로 변경

JoSH0318 committed:
[rename]: modify method, parameter, variable name

+ 쓸모없는 메서드 제거
+ 공백제거

marisol-develop committed:
[correct]: modify closer parameter

- forEach문 내의 $0 -> fruit으로 변경

marisol-develop committed:
Merge pull request #5 from JoSH0318/5_mmim90-3

5 mmim90 3

JoSH0318 committed:
[docs]: update README.md

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -29,31 +29,24 @@ final class OrderViewController: UIViewController {
             case .mango:
-                mangoStockLabel.text = convertFruitStockToString(.mango)
+                mangoStockLabel.text = juiceMaker.fruitStore.inventory[.mango]?.description
             }
```
반복문 안에서 switch 문을 또 사용하셨네요!

이렇게 되면 2중 반복문을 사용하신 것과 같아요.

지금 전반적으로 이렇게 쓰인 코드들이 많은데 어떻게 하면 개선할 수 있을지 고민해주세요!

marisol-develop replied:
Fruit 열거형에 rawValue를 줘서 stockLabels의 index로 활용했습니다.
반복되는 코드들은 많이 줄었는데, 아래 고민한점 코멘트 참고 부탁드립니다!🙇‍♀️

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -62,29 +55,26 @@ final class OrderViewController: UIViewController {
         }
-        stockViewController.modalTransitionStyle = .coverVertical
-        stockViewController.modalPresentationStyle = .automatic
+        stockViewController.delegate = self
+        stockViewController.currentStocks = juiceMaker.fruitStore.inventory
```
Order -> Stock은 직접 전달을 하셨고
Stock -> Order은 델리게이트로 전달하셨는데

각각 장단점은 무엇이 있을까요??

JoSH0318 replied:
property 직접 전달의 장점은 간단하다 라는 것 입니다. 특별한 패턴 같은게 존재하지 않아 코드가 짧고 쉽습니다. 단점은 객체간의 결합도가 높아지는 것 같습니다. 다른 class의 property를 직접 전달하다보니 서로 의존하게 된다고 생각했습니다. 또한, push/present 방식의 화면전환에서만 쓰일 수 있다는 것도 단점인 것 같습니다.

delegate로 전달하는 것의 장점은 객체간 서로 의존하지 않아 결합도가 낮아진다는 것 입니다. 단점으로는 특별한 패턴이 필요하여 코드가 길어진다는 것, delegate에 nil 값이 들어가지 않도록 주의해야한다는 점이 있는 것 같습니다.

저희가 알고 있는 장단점은 여기까지 입니다. 혹시 저희가 잘못 생각하고 있거나 좀 더 알아야할 특징이 있다면 알려주세요 😍

stevenkim18 replied:
잘 파악하셨습니다!! 여기에도 정답은 없습니다. 때에 따라 적절하게 사용하시면 됩니다.
객체간의 결합도를 낮추려면 특정 기술이 필요하고 그로 인해서 코듸 가독성이 낮아집니다.
delegate도 마찬가지로 notification center보다는 덜 하지만 이 또한 한눈에 관련 로직을 알아보기는 어렵다고 저는 생각됩니다.

계속 객체 설계하시는 것을 많이 고민해보시고 연습해 보시면 정말 많이 도움이 되실꺼에요!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -62,29 +55,26 @@ final class OrderViewController: UIViewController {
-        default:
-            alertUnknownError()
         }
         showCurrentStock()
```
`order` 함수가 반복됩니다. 어떻게 줄일 수 있을까요??

marisol-develop replied:
sender의 tag와 Juice 열거형 case의 rawValue가 같은 경우 order 메서드에 넣을 juice 파라미터를 받아서 중복 코드를 줄이는 식으로 구현했습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/StockViewController.swift:
```diff
@@ -7,14 +7,72 @@
+                mangoStockLabel.text = currentStocks[.mango]?.description
+                mangoStepper.value = Double(currentStocks[.mango] ?? 0)
+            }
+        }
```
`currentStocks[fruit]`이 계속 반복 되네요!!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/StockViewController.swift:
```diff
@@ -7,14 +7,72 @@
+            mangoStockLabel.text = Int(sender.value).description
+            currentStocks[.mango] = Int(sender.value)
+        }
     }
```
`Int(sender.value)` 가 반복되네요!!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -95,9 +85,7 @@ final class OrderViewController: UIViewController {
+        } catch { }
     }
 }
 
```
alert 관련 함수를 extension으로 빼신 것은 정말 잘하셨습니다.
`//Mark: Alert` 같이 마크 주석으로 표시 하면 더 좋을 것 같아요!!

stevenkim18 reviewed (commented):
밈 마리솔 마지막 스탭까지 너무 고생많으셨습니다!!

3주동안 배우신 것을 한번 쭉 복습해 보시기를 권장드립니다!!

너무너무 중요하고 기본 중에 기본이 많이 담긴 프로젝트 입니다!!

### ****Switch문 & if문****

- 코드가 무결점하는 것이 정확히 어떤 의미인가요?
- 저의 개인적인 생각으로는 if - else 문은 확실히 else까지 정의해줘야 한다고 생각합니다.
- 오히려 저는 Switch 문을 쓰고

```swift
defalut:
	return 
```

이런식으로 처리해 줄 것 같아요.

### ****Dictionary Optional Binding****

- `??` 를 사용하는 것은 안전한 바인딩입니다.
- `??` 를 사용하고 싶지 않으시다면 따로 재고를 조회하는 함수를 만들어 줄 수 있을 것 같아요

```swift
private stock(of: friut Friut) -> Int {
	return currentStocks[fruit] ?? 0
}
```

### ****Alert 합치기****

```swift
func alert(title: String, message: String, actions: [UIAlertAction])
```

- 저라면 이렇게 할 것 같아요!

### ****Auto-Layout과 UIStepper****

- 네네 모든 UI는 커스텀이 가능합니다
- 상속받아서 커스텀 할 수 있고 만약 되지 않는다면 `CA` 가 붙은 더 low 단 코드로 커스텀 할 수 있습니다!

### ****Property observer****

- 네네 필요에 따라서 많이 사용합니다. 저는 사용해본적은 없지만 예시로 주신 코드 처럼 사용한 예제를 많이 보았습니다.
- 앞으로 많이 고민하셔야 할 부분입니다! 이렇게 하는 것을 데이터 바이딩이라고 하는데 didset, 델리게이터, Notification Center, 직접 연결등 정말 다양합니다!!

marisol-develop committed:
[improve]: modify orderButtonsClicked() logic

- if - else if문을 tag와 rawValue를 사용하여 수정
- Juice 열거형에 CaseIterable 프로토콜 채택

marisol-develop committed:
Merge branch '5_mmim90-3'

JoSH0318 committed:
[improve]: modify method's logics

- showCurrentStock, stepperClicked 메서드 로직 변경
- UILabel을 Outlet Collection 묶음
- UIStepper을 Outlet Collection 묶음

marisol-develop committed:
[improve]: modify showCurrentStock() logic

- forEach문으로 메서드 로직 수정
- 불필요한 공백 제거
- Alert 마크 주석 표시

JoSH0318 committed:
[docs]: update README.md

- UML
- 고민한점
- 배운개념

JoSH0318 commented:
@stevenkim18  스티븐! 코멘트 내용을 기반으로 수정했습니다! 

### 고민한점
#### 코드의 간결성 vs 코드의 가독성
- 반복되는 코드를 줄이기 위해 label과 stepper를 outletCollection으로 묶어주었습니다. 그리고 과일의 재고를 표시하는 라벨마다 tag를 붙이고, Fruit 열거형의 rawValue에 원시값을 넣어 tag를 이용해 라벨의 outletCollection에 index로 접근했습니다.

```swift
Fruit.allCases.forEach{ fruit in
    stockLabels[fruit.rawValue].text = currentStocks[fruit]?.description
    stockSteppers[fruit.rawValue].value = Double(currentStocks[fruit] ?? 0)
}
```

이렇게 하면 반복되는 코드가 줄고 매우 간결하게 표현할 수 있다는 장점이 있지만, 과일이 추가될 때마다 tag를 달아줘야하고, 혹시라도 tag과 fruit의 rawValue가 매칭되지 않는 경우에 문제가 발생할 수 있다는 단점이 있는 것 같습니다. 이전 코드보다 매우 간결해졌지만 가독성 부분에서 좋지 않은 것 같다는 생각도 들었습니다.

#### UML 작성법
UML 작성법을 학습하고 학습한 내용대로 작성해보았습니다 🤗
아직 UML을 통해 소통하는 것이 미흡한 것 같습니다. 
특히나 화살표 표기가 어렵고, UML 작성법이 JAVA를 기준으로 하다보니 swift에서의 작성법이 헷갈립니다.
 부족한 부분이나 이해가 되지 않는 부분을 말씀해주시면 감사하겠습니다! 🙏
<img width="4612" alt="JuiceMaker_UML_STEP3" src="https://user-images.githubusercontent.com/88810018/157646900-0be73801-08e4-4ebb-851e-b86abf9012d7.png">

marisol-develop committed:
[docs]: update README.md

- 프로젝트 소개 및 기능
- 기술의 장단점

stevenkim18 commented:
### 코드의 간결성 vs 코드의 가독성
좋은 고민 포인트입니다. 
아쉽지만 이부분도 정답은 없습니다 ㅜㅜ
지금 작성하신 코드에서 그나마 가독성을 높이려면 저는 `fruit.rawValue` -> `fruit.tag` 라는 프로퍼티를 만들어줘서 의미를 더 명확하게 할 것 같아요.

최고의 코드는 없지만 최선의 코드는 있다고 생각합니다. 위와 같은 고민은 정말 계속 해야되는 것 같습니다!

### UML
uml을 배우셨군요!!
자바와 스위프트라고 해서 크게 다를 것은 없습니다. 자바의 인터페이스가 swift에서는 프로토콜로 생각하시면 됩니다.

저는 아직 uml을 많이 사용해본적이 없지만
저라면 다른 분들이 보기 좋게 도메인 별로나 역활 별로 조금 구분을 지을 것 같습니다. 오히려 객체들이 많아지면 더 알아보기 어려울 것 같아요.

지금 보내주신 것에서는 메인과 재고 뷰컨을 조금 멀리 분리해두고 각 관련된 객체들을 정렬할 것 같습니다.

화살표는 잘 사용하셨습니다! 

---

밈 마리솔 3주동안 너무 고생많으셨습니다. 👏👏👏👏👏👏👏👏👏
디엠으로 요청드린 리드미는 꼭 작성해주세요!! 리뷰는 끝났지만 나중에 또 궁금하신 것 있으시면 언제든지 질문 주세요!!
